force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,41 +227,67 @@ public with sharing class DMLOperations {
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
+            public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+                
+                List<Contact> contacts = new List<Contact>();
+                // Loop through each Contact
+                for(Contact contact : contactsToUpsert){
+                // Extract the Account Name from Contact's LastName
+                String accountName = contact.LastName;
 
-		// Extract the Account Name from Contact's LastName
+                // Get the account based on the name or create one if it doesn't exist
+                    Account account = upsertAccount(accountName);
```

Comment:
Nice job using taking advantage of the **upsertAccount()** method. (Too many students have ignored that hint!)